Enabling auto-compiled models on EF 9.0.0-rc.2 causes syntax errors on build/publish

I'm testing the new EF Core 9 feature [auto-compiled models](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-9.0/whatsnew#auto-compiled-models) right now, and have noticed that enabling it causes build failures due to syntax errors in generated code (files suffixed `EntityType.g.cs`).

Example errors observed:

```
.../net9.0/EmojiEntityType.g.cs(142,32): error CS1026: ) expected
.../net9.0/AntennaEntityType.g.cs(713,25): error CS1026: ) expected
.../net9.0/EmojiEntityType.g.cs(143,135): error CS1002: ; expected
.../net9.0/AntennaEntityType.g.cs(714,123): error CS1002: ; expected
.../net9.0/AntennaEntityType.g.cs(714,123): error CS1513: } expected
.../net9.0/EmojiEntityType.g.cs(143,135): error CS1513: } expected
.../net9.0/FilterEntityType.g.cs(197,117): error CS1026: ) expected
.../net9.0/FilterEntityType.g.cs(208,27): error CS1002: ; expected
.../net9.0/FilterEntityType.g.cs(208,27): error CS1513: } expected
.../net9.0/OauthAppEntityType.g.cs(327,24): error CS1026: ) expected
.../net9.0/MessagingMessageEntityType.g.cs(331,23): error CS1026: ) expected
.../net9.0/MessagingMessageEntityType.g.cs(332,123): error CS1002: ; expected
.../net9.0/OauthAppEntityType.g.cs(328,137): error CS1002: ; expected
.../net9.0/MessagingMessageEntityType.g.cs(332,123): error CS1513: } expected
.../net9.0/OauthAppEntityType.g.cs(328,137): error CS1513: } expected
.../net9.0/OauthAppEntityType.g.cs(391,23): error CS1026: ) expected
.../net9.0/GalleryPostEntityType.g.cs(236,27): error CS1026: ) expected
.../net9.0/OauthAppEntityType.g.cs(392,125): error CS1002: ; expected
.../net9.0/OauthAppEntityType.g.cs(392,125): error CS1513: } expected
.../net9.0/GalleryPostEntityType.g.cs(237,131): error CS1002: ; expected
.../net9.0/GalleryPostEntityType.g.cs(237,131): error CS1513: } expected
.../net9.0/MetaEntityType.g.cs(141,32): error CS1026: ) expected
.../net9.0/MetaEntityType.g.cs(142,145): error CS1002: ; expected
.../net9.0/GalleryPostEntityType.g.cs(394,24): error CS1026: ) expected
.../net9.0/MetaEntityType.g.cs(142,145): error CS1513: } expected
.../net9.0/GalleryPostEntityType.g.cs(395,121): error CS1002: ; expected
.../net9.0/GalleryPostEntityType.g.cs(395,121): error CS1513: } expected
```

Affected project/branch: https://iceshrimp.dev/iceshrimp/Iceshrimp.NET/src/branch/wip/net90-optimize2 (error occurs when building the `Iceshrimp.Backend` subfolder).

I'm not sure whether this is a bug in EF, or a bug with Npgsql. I've tested it against EF versions `9.0.0-rc.2`, `9.0.0` (GA), as well as various `rtm` nightly builds. The issue also occurs against efcore.pg version `9.0.0-rc.2-ci.20241106T153535`.

Any input would be greatly appreciated!

Yep, that's it. Minimally reproducible example:

```
dotnet new webapi -o api
cd api
```

`DbContext.cs`
```csharp
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

namespace api;

public class DatabaseContext(DbContextOptions<DatabaseContext> options) : DbContext(options)
{
	public virtual DbSet<Example> Objects { get; init; } = null!;

	protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) {
		optionsBuilder.UseNpgsql();
	}

	protected override void OnModelCreating(ModelBuilder modelBuilder) {
		modelBuilder.Entity<Example>().HasKey(p => p.Id);
	}
}

public class Example
{
	[Key]
	[Column("id", TypeName = "character varying(32)")]
	[StringLength(32)]
	public string Id { get; set; } = null!;

	[Column("test", TypeName = "character varying(32)[]")]
	public List<string> Test { get; set; } = [];
}
```

`api.csproj`
```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

    <PropertyGroup>
        <TargetFramework>net9.0</TargetFramework>
        <Nullable>enable</Nullable>
        <ImplicitUsings>enable</ImplicitUsings>
        <EFOptimizeContext>true</EFOptimizeContext>
        <EFPrecompileQueriesStage>none</EFPrecompileQueriesStage>
    </PropertyGroup>

    <ItemGroup>
        <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="9.0.0-rc.2" />
        <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="9.0.0" />
        <PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="9.0.0" />
        <PackageReference Include="Microsoft.EntityFrameworkCore" Version="9.0.0" />
        <PackageReference Include="Microsoft.EntityFrameworkCore.Tasks" Version="9.0.0" />
        <PackageReference Include="Microsoft.AspNetCore.DataProtection.EntityFrameworkCore" Version="9.0.0" />
    </ItemGroup>

</Project>
```

`Program.cs`
```csharp
// [...]
builder.Services.AddNpgsql<DatabaseContext>(null);
// [...]
```

```
> dotnet publish
.../obj/Release/net9.0/ExampleEntityType.g.cs(140,31): error CS1026: ) expected
.../obj/Release/net9.0/ExampleEntityType.g.cs(141,129): error CS1002: ; expected
.../obj/Release/net9.0/ExampleEntityType.g.cs(141,129): error CS1513: } expected
```

--- a/src/EFCore.PG/Design/Internal/NpgsqlCSharpRuntimeAnnotationCodeGenerator.cs
+++ b/src/EFCore.PG/Design/Internal/NpgsqlCSharpRuntimeAnnotationCodeGenerator.cs
@@ -14,8 +14,11 @@ namespace Npgsql.EntityFrameworkCore.PostgreSQL.Design.Internal;
 ///     doing so can result in application failures when updating to a new Entity Framework Core release.
 /// </summary>
 #pragma warning disable EF1001 // Internal EF Core API usage.
-public class NpgsqlCSharpRuntimeAnnotationCodeGenerator : RelationalCSharpRuntimeAnnotationCodeGenerator
+public class NpgsqlCSharpRuntimeAnnotationCodeGenerator
+    : RelationalCSharpRuntimeAnnotationCodeGenerator, ICSharpRuntimeAnnotationCodeGenerator
 {
+    private int _typeMappingNestingCount;
+
     /// <summary>
     ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
     ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
@@ -42,8 +45,31 @@ public class NpgsqlCSharpRuntimeAnnotationCodeGenerator : RelationalCSharpRuntim
         ValueComparer? keyValueComparer = null,
         ValueComparer? providerValueComparer = null)
     {
-        var result = base.Create(typeMapping, parameters, valueComparer, keyValueComparer, providerValueComparer);
+        _typeMappingNestingCount++;
 
+        try
+        {
+            var result = base.Create(typeMapping, parameters, valueComparer, keyValueComparer, providerValueComparer);
+            AddNpgsqlTypeMappingTweaks(typeMapping, parameters);
+            return result;
+        }
+        finally
+        {
+            _typeMappingNestingCount--;
+        }
+    }
+
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    [EntityFrameworkInternal]
+    protected virtual void AddNpgsqlTypeMappingTweaks(
+        CoreTypeMapping typeMapping,
+        CSharpRuntimeAnnotationCodeGeneratorParameters parameters)
+    {
         var mainBuilder = parameters.MainBuilder;
 
         var npgsqlDbTypeBasedDefaultInstance = typeMapping switch
@@ -57,6 +83,8 @@ public class NpgsqlCSharpRuntimeAnnotationCodeGenerator : RelationalCSharpRuntim
 
         if (npgsqlDbTypeBasedDefaultInstance is not null)
         {
+            CheckElementTypeMapping();
+
             var npgsqlDbType = ((INpgsqlTypeMapping)typeMapping).NpgsqlDbType;
 
             if (npgsqlDbType != npgsqlDbTypeBasedDefaultInstance.NpgsqlDbType)
@@ -83,6 +111,8 @@ public class NpgsqlCSharpRuntimeAnnotationCodeGenerator : RelationalCSharpRuntim
         switch (typeMapping)
         {
             case NpgsqlEnumTypeMapping enumTypeMapping:
+                CheckElementTypeMapping();
+
                 var code = Dependencies.CSharpHelper;
                 mainBuilder.AppendLine(";");
 
@@ -119,6 +149,8 @@ public class NpgsqlCSharpRuntimeAnnotationCodeGenerator : RelationalCSharpRuntim
 
             case NpgsqlRangeTypeMapping rangeTypeMapping:
             {
+                CheckElementTypeMapping();
+
                 var defaultInstance = NpgsqlRangeTypeMapping.Default;
 
                 var npgsqlDbTypeDifferent = rangeTypeMapping.NpgsqlDbType != defaultInstance.NpgsqlDbType;
@@ -154,7 +186,14 @@ public class NpgsqlCSharpRuntimeAnnotationCodeGenerator : RelationalCSharpRuntim
             }
         }
 
-        return result;
+        void CheckElementTypeMapping()
+        {
+            if (_typeMappingNestingCount > 1)
+            {
+                throw new NotSupportedException(
+                    $"Non-default Npgsql type mappings ('{typeMapping.GetType().Name}' with store type '{(typeMapping as RelationalTypeMapping)?.StoreType}') aren't currently supported as element type mappings, see https://github.com/npgsql/efcore.pg/issues/3366.");
+            }
+        }
     }
 
     /// <inheritdoc />

